### PR TITLE
Dpruitt/zacc layers port

### PR DESCRIPTION
**Pull request by daviddpruitt**
# Modulus Pull Request

## Description
Add features to support running more sophisticated ocean models

- Bounded Constraint
- Initial convolutional block

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py [hunk 1]
```diff
@@ -67,7 +67,10 @@ def __init__(
             self.channels = out_channels
         else:
             self.channels = in_channels
-        self.scaling = scaling
+        if scaling is not None:
+            self.scaling = scaling
+        else:
```
Comment (yikwill) [inline physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:72]: I think we should still explicitly require a scaling dict or at least throw a clear warning if it is not provided. This leaves the potential for silently falling back to 0 mean 1 std scaling which could be disastrous depending on the variable.

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py [hunk 2]
```diff
@@ -102,6 +105,100 @@ def forward(self, prediction, input):
...
+        super().__init__()
+        self.bounds = bounds
+        if out_channels is not None:
+            self.channels = out_channels
+        else:
+            self.channels = in_channels
+        if scaling is not None:
+            self.scaling = scaling
+        else:
+            self.scaling = {var: {'mean': 0.0, 'std': 1.0} for var in self.variables}
```
Comment (yikwill) [inline physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:145]: This else case will fail since `self.variables` is not defined yet.

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py [hunk 3]
```diff
@@ -102,6 +105,100 @@ def forward(self, prediction, input):
...
+        max_per_channel = []
+        for name in self.channels:
+            phys_min, phys_max = (
+                self.bounds[name] if name in constrained_set else (None, None)
+            )
+            if phys_min is None:
+                min_per_channel.append(float("-inf"))
+            else:
+                min_per_channel.append(
+                    (phys_min - scaling[name]["mean"]) / scaling[name]["std"]
```
Comment (yikwill) [inline physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:171]: Should we be using `self.scaling` here as well as in `NonnegativeConstraint`?

File: test/models/dlwp_healpix/test_healpix_constraints.py
Comment (yikwill) [inline test/models/dlwp_healpix/test_healpix_constraints.py:1]: If we decide to keep `scaling` as an optional argument, can we add a test for when it is not provided?

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py [hunk 4]
```diff
@@ -102,6 +105,100 @@ def forward(self, prediction, input):
             return replace_value_keep_gradient(prediction, clamped)
         return clamped
 
+class BoundConstraint(torch.nn.Module):
+    def __init__(
+        self,
+        bounds: dict[str, list[float]],
```
Comment (yikwill) [inline physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:111]: Nit: this type hint doesn't allow for the `None` that the docstring describes

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py [hunk 5]
```diff
@@ -102,6 +105,100 @@ def forward(self, prediction, input):
...
+        # Only apply constraint to variables that are used by model
+        missing = [var for var in self.bounds if var not in self.channels]
+        self.variables = [var for var in self.bounds if var in self.channels]
+
+        if missing:
+            logger0.warning(
+                f"Requested bound constrained variables "
+                f"{missing} not found in model channels and will be ignored."
+            )
+
```
Comment (yikwill) [inline physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:157]: I think a check that the bound min is less than the bound max would be good to add. Otherwise, `torch.clamp` can silently make everything equal to `max`.

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py [hunk 6]
```diff
@@ -102,6 +105,100 @@ def forward(self, prediction, input):
...
+            if the clamp were identity so out-of-bounds predictions still get a
+            learning signal. Default False.
+        """
+        super().__init__()
+        self.bounds = bounds
+        if out_channels is not None:
+            self.channels = out_channels
+        else:
+            self.channels = in_channels
+        if scaling is not None:
```
Comment (yikwill) [inline physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:142]: We should have a check (like the one in `NonnegativeConstraint`) that ensures that each constrained variable is in the scaling dict.

**Review by yikwill — COMMENTED** [review]
I think we should decide whether `scaling` should be a required argument for the constraints or not. I personally think it should be, but if we don't require it we should at least have a warning to avoid a silent fallback to 0 mean 1 std scaling. Besides that, a few small comments.

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py [hunk 7]
```diff
@@ -40,8 +40,8 @@ def __init__(
         self,
         variables: list[str],
         in_channels: list[str],
```
Comment (smartin98) [inline physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:42]: I notice some of the ocean checkpoint configs use 'channels' rather than 'in_channels' for the bound constraint. Should we add backwards compatibility here? Something like:

```python
def __init__(
    self,
    bounds: dict[str, list[float]],
    in_channels: list[str] = None,
    out_channels: list[str] = None,
    scaling: dict[str, dict[str, float]] = None,
    keep_grad_through_clamp: bool = False,
    channels: list[str] = None,
):
    if in_channels is None and channels is not None:
        logger0.warning(
            "channels is deprecated, use in_channels instead"
        )
        in_channels = channels
    if in_channels is None:
        raise TypeError("BoundConstraint requires in_channels (or legacy channels)")
```

**Comment by smartin98** [comment]
The other issue that came up for me inferencing the ocean checkpoints from this branch was some dtype mismatches:

https://github.com/AtmosSci-DLESM/modulus-uw/blob/5506fa2f643a54211ec8ea7c1c750bd14e630a05/physicsnemo/models/dlwp_healpix_layers/healpix_paddings.py#L769

I had to append .long() after the .to() call. 

Same for pos_v1 here: https://github.com/AtmosSci-DLESM/modulus-uw/blob/5506fa2f643a54211ec8ea7c1c750bd14e630a05/physicsnemo/models/dlwp_healpix_layers/healpix_paddings.py#L774

and for i1_sub here: https://github.com/AtmosSci-DLESM/modulus-uw/blob/5506fa2f643a54211ec8ea7c1c750bd14e630a05/physicsnemo/models/dlwp_healpix_layers/healpix_paddings.py#L777

**Comment by smartin98** [comment]
@daviddpruitt I think the source of the flashing ocean predictions where day 4 and day 8 preds oscillate comes from this edit to the forward pass Zac has in his source branch. Basically the issue is since the model predicts residual and has 2-in-2-out structure, it matters what persistence state you predict the residual around. The default was copying 2 prior steps and adding residual on top, so day 4 and day 8 residuals being added on top of different background states. 

https://github.com/AtmosSci-DLESM/modulus-uw/blob/5506fa2f643a54211ec8ea7c1c750bd14e630a05/physicsnemo/models/dlwp_healpix/HEALPixUNet.py#L440

The above line should be replaced by either 'Option 2' or 'Option 3' below:

Original Zac implementation:
https://github.com/AtmosSci-DLESM/modulus-uw/blob/5a6e87ae063c187af85fc34d387d796e832561ed/physicsnemo/models/dlwp_healpix/HEALPixUNet.py#L445-L481

Option 2 has same inference cost as Option 1 while getting rid of the flashing issue (by persisting most recent IC as the background state). Option 3 looks like the most physically sensible but requires an extra for loop (not that expensive for 2-in-2-out though); in this one the day 4 pred is taken as the background for the day 8 residual. Option 3 can probably be implemented more efficiently with a cumsum operation rather than for loop.

**Comment by smartin98** [comment]
> @daviddpruitt I think the source of the flashing ocean predictions where day 4 and day 8 preds oscillate comes from this edit to the forward pass Zac has in his source branch. Basically the issue is since the model predicts residual and has 2-in-2-out structure, it matters what persistence state you predict the residual around. The default was copying 2 prior steps and adding residual on top, so day 4 and day 8 residuals being added on top of different background states.
> 
> https://github.com/AtmosSci-DLESM/modulus-uw/blob/5506fa2f643a54211ec8ea7c1c750bd14e630a05/physicsnemo/models/dlwp_healpix/HEALPixUNet.py#L440
> 
> The above line should be replaced by either 'Option 2' or 'Option 3' below:
> 
> Original Zac implementation:
> 
> https://github.com/AtmosSci-DLESM/modulus-uw/blob/5a6e87ae063c187af85fc34d387d796e832561ed/physicsnemo/models/dlwp_healpix/HEALPixUNet.py#L445-L481
> 
> Option 2 has same inference cost as Option 1 while getting rid of the flashing issue (by persisting most recent IC as the background state). Option 3 looks like the most physically sensible but requires an extra for loop (not that expensive for 2-in-2-out though); in this one the day 4 pred is taken as the background for the day 8 residual. Option 3 can probably be implemented more efficiently with a cumsum operation rather than for loop.

Pretty sure Option 2 is what Zac's checkpoints were trained with so let's go with that.

File: physicsnemo/models/dlwp_healpix/HEALPixRecUNet.py [hunk 8]
```diff
@@ -174,6 +186,13 @@ def __init__(
         self.compile_padding = compile_padding
         self.nside = nside
 
+        if isinstance(self.residual_prediction, bool):
+            legacy_residual_prediction = self.residual_prediction
+            self.residual_prediction = "same-index-add" if legacy_residual_prediction else "none"
+            logger.warning(f"Residual prediction type {legacy_residual_prediction} is deprecated. Using '{self.residual_prediction}' instead. Please update your configuration.")
+        if self.residual_prediction not in ["none", "same-index-add", "last-reference", "chained"]:
+            raise ValueError(f"Invalid residual prediction type: {self.residual_prediction}")
+
```
Comment (smartin98) [inline physicsnemo/models/dlwp_healpix/HEALPixRecUNet.py]: Is there any reason to keep supporting the same-index-add option? To my mind this option is essentially a bug that should not be supported. Zac's checkpoints were trained assuming the "last-reference" option, so unless you have ocean checkpoints that assume this option, or unless this is relevant for the atmos model, my vote would be to reduce confusion by dropping "same-index-add" option.

Comment (smartin98) [reply physicsnemo/models/dlwp_healpix/HEALPixRecUNet.py]: Perhaps others disagree though? @yikwill @akshaysubr for viz.

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py (hunk 5 above)
Comment (daviddpruitt) [reply physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:157]: Bounds check added. If min > max an error is thrown.

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py (hunk 4 above)
Comment (daviddpruitt) [reply physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:111]: Fixed the type hint

File: test/models/dlwp_healpix/test_healpix_constraints.py
Comment (daviddpruitt) [reply test/models/dlwp_healpix/test_healpix_constraints.py:1]: Test added

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py (hunk 3 above)
Comment (daviddpruitt) [reply physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:171]: Scaling for all constraints is switched to `scaling`. The scaling values are only needed for setting the bounds so they don't need to be kept around.

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py (hunk 2 above)
Comment (daviddpruitt) [reply physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:145]: Fixed.

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py (hunk 1 above)
Comment (daviddpruitt) [reply physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:72]: If scaling isn't provided a warning is triggered now.

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py (hunk 6 above)
Comment (daviddpruitt) [reply physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:142]: Added check for scaling

File: physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py (hunk 7 above)
Comment (daviddpruitt) [reply physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:42]: I added the backwards compatibility option. Although I'm wondering if it would be better to go with a single `channels` argument since there isn't any need to differentiate between `in_channels` and `out_channels`

**Review by yikwill — APPROVED** [review]
LGTM once the backwards compatible `channels` argument is added to the `BoundConstraint`. I think Scott's issues with dtype in the isolatitude padding layers can be handled in another PR since I don't believe it's an issue related to the layers this PR is adding.
